05/nicdard.ml

```diff
@@ -0,0 +1,83 @@
+  ; "subtype_subrstruct_sub1", passtest @@ Typechecker.subtype (struct_ctxt s3 s2) (TRef (RStruct (fst s2))) (TRef (RStruct (fst s3)))
+  ; "subtype_subrstruct_no_sub1", failtest @@ Typechecker.subtype (struct_ctxt s1 s6) (TRef (RStruct (fst s1))) (TRef (RStruct (fst s6)))
+  ; "subtype_subrstruct_no_sub2", failtest @@ Typechecker.subtype (struct_ctxt s3 s4) (TRef (RStruct (fst s4))) (TRef (RStruct (fst s3)))
+  ; "subtype_subrstruct_no_sub3", failtest @@ Typechecker.subtype (struct_ctxt s5 s4) (TRef (RStruct (fst s4))) (TRef (RStruct (fst s5)))
+  ; "subtype_subrfunt_eq", passtest @@ Typechecker.subtype Tctxt.empty (TRef (RFun ([TInt], RetVal TInt))) (TRef (RFun ([TInt], RetVal TInt)))
+  ; "subtype_subrfunt_controvariant_void", passtest @@ Typechecker.subtype Tctxt.empty (TRef (RFun ([TNullRef RString], RetVoid))) (TRef (RFun ([TRef RString], RetVoid)))
+  ; "subtype_subrfunt_covariant_void", failtest @@ Typechecker.subtype Tctxt.empty (TRef (RFun ([TRef RString], RetVoid))) (TRef (RFun ([TNullRef RString], RetVoid)))
+  ; "subtype_subrfunt_contro_covariant", passtest @@ Typechecker.subtype Tctxt.empty (TRef (RFun ([TNullRef RString], RetVal (TRef RString)))) (TRef (RFun ([TRef RString], RetVal (TNullRef RString))))
+  ; "subtype_subrfunt_contro_contro", failtest @@ Typechecker.subtype Tctxt.empty (TRef (RFun ([TNullRef RString], RetVal (TNullRef RString)))) (TRef (RFun ([TRef RString], RetVal (TRef RString))))
+  ]
+
+let typecheck_ty_unit_tests =
+  let passtest (f : assertion) = (fun () -> try f () with Typechecker.TypeError error -> failwith @@ "Type error: " ^ error) in
+  let failtest (f : assertion) = (fun () -> try f (); failwith "Should have raised TypeError." with Typechecker.TypeError error -> ()) in
+  [ "wf_typokokint", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty TInt)
+  ; "wf_typokokbool", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty TBool)
+  ; "wf_typokokreft", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TRef RString)) 
+  ; "wf_typokokreftq", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TNullRef RString))
+  ; "array_of_undef_struct", failtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TNullRef (RArray (TRef (RStruct "Color")))))
+  ; "array_of_def_struct", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) (Tctxt.add_struct Tctxt.empty "Color" []) (TNullRef (RArray (TRef (RStruct "Color")))))
+  ; "fun_undef_struct", failtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TNullRef (RFun ([ TRef (RStruct "Color"); TInt; TBool ], RetVoid))))
+  ; "fun_def_struct", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) (Tctxt.add_struct Tctxt.empty "Color" []) ((TNullRef (RFun ([ TRef (RStruct "Color"); TInt; TBool ], RetVal TInt)))))
+  ]
+
+let tctxt path (builder : Typechecker.context_builder) =
```

Comment:
Typechecker.context_builder is not defined in the original typechecker.ml

Reply:
You are right sorry! I added a comment with the required type definition to be added to use them, otherwise I can remove them

Reply:
The comment is in the next commit

Reply:
I'd propose to replace 'Typechecker.context_builder' with the actual type to safe extra editing work. 
But I leave this up to you to decide else LGTM.

Reply:
Cool thanks!